### ISSUE TITLE
Train T5 with adapter without using Trainer Class didn't update the adapter's weight

### ISSUE THREAD

**Issue by mistel1225** [Adapter-Hub/adapter-transformers#260]
## Environment info

- Platform: Linux-5.11.0-41-generic-x86_64-with-glibc2.29
- Python version: 3.8.10
- PyTorch version (GPU?): 1.10.0+cu113 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: <fill in>
- Using distributed or parallel set-up in script?: <fill in>
     

## Details
Hi, I tried to train adapter with T5 and noticed that the loss is really high,
I try to print the adapter weight in the first layer in encoder and it seems that the weight didn't change during training,
I've read the suggestion from [docs](https://github.com/Adapter-Hub/adapter-transformers/blob/master/adapter_docs/training.md) and #231 , nothing find.
Here is my training script:

```
import torch
import logging
import os
import sys
import numpy as np
from torch.utils.data import Dataset, DataLoader
from adapterT5_Dataset import getDataset
from transformers import (
        AdamW, 
        Adafactor, 
        T5ForConditionalGeneration, 
        T5Tokenizer, 
        HoulsbyConfig, 
        EvalPrediction,
        get_scheduler,
        set_seed,
        default_data_collator)
from adapterT5_Dataset import getDataset
import random
import math
from tqdm import tqdm
from tqdmHandler import tqdmLoggingHandler
logger = logging.getLogger(__name__)
def train_adapterT5(args, compute_metric):
    logging.basicConfig(
       	format="%(asctime)s - %(levelname)s - %(name)s - %(message)s",
        datefmt="%m/%d/%Y %H:%M:%S",
        level=logging.INFO,
        handlers=[tqdmLoggingHandler()],
        )
    set_seed(args.seed)
    train_dataset = getDataset(args, args.task_name+'/train.csv')
    val_dataset = getDataset(args, args.task_name+'/val.csv')
    train_dataloader = DataLoader(train_dataset, shuffle=True, batch_size=args.train_batch_size)
    val_dataloader = DataLoader(val_dataset, batch_size=args.eval_batch_size)
    model = T5ForConditionalGeneration.from_pretrained(args.model_name)
    no_decay = ["bias", "LayerNorm.weight"]
    optimizer_grouped_parameters = [
        {
            "params": [p for n, p in model.named_parameters() if not any(nd in n for nd in no_decay)],
            "weight_decay": args.weight_decay,
        },
        {
            "params": [p for n, p in model.named_parameters() if any(nd in n for nd in no_decay)],
            "weight_decay": 0.0,
        },
    ]
    optimizer = AdamW(optimizer_grouped_parameters, lr=args.lr, eps=args.adam_epsilon)
    #model = T5ForConditionalGeneration.from_pretrained(args.model_name)
    #tokenizer = T5Tokenizer.from_pretrained(args.tokenizer_name)
    num_update_steps_per_epoch = math.ceil(len(train_dataloader) / args.gradient_accumulation_steps)
    max_train_steps = args.epoch * num_update_steps_per_epoch
    lr_scheduler = get_scheduler(
        name=args.lr_scheduler_type,
        optimizer=optimizer,
        num_warmup_steps=args.num_warmup_steps,
        num_training_steps=max_train_steps
            )
    #model = T5ForConditionalGeneration.from_pretrained(args.model_name).to(args.device)
    tokenizer = T5Tokenizer.from_pretrained(args.tokenizer_name)
    adapter_config = HoulsbyConfig()
    model.add_adapter(args.task_name, config=adapter_config)
    model.to(args.device)
    print(model)
    progress_bar = tqdm(range(max_train_steps))
    completed_steps = 0
    metric_list = {}
    model.train_adapter(args.task_name)
    model.set_active_adapters(args.task_name)
    #model.freeze_model(False)
    for epoch in range(args.epoch):
        metric_list[epoch] = {'train_loss':[]}
        #model.train_adapter(args.task_name)
        #model.train()
        for step, batch in enumerate(train_dataloader):
            model.train()
            #model.train_adapter(args.task_name)
            #model.set_active_adapters(args.task_name)
            outputs = model(input_ids=batch['source_ids'].to(args.device),
                    attention_mask=batch['source_mask'].to(args.device),
                    labels=batch['target_ids'].to(args.device),
                    decoder_attention_mask=batch["target_mask"].to(args.device)
                    )
            loss = outputs.loss
            #metric_list[epoch]['train_loss'].append(loss.item())
            loss = loss / args.gradient_accumulation_steps
            loss.backward()
            #print(loss)
            if step % args.gradient_accumulation_steps == 0 or step == len(train_dataloader) -1 :
                optimizer.step()
                lr_scheduler.step()
                optimizer.zero_grad()
                progress_bar.update(1)
                completed_steps += 1
                #print(model.encoder.block[0].layer[1].adapters.rotten_tomatoes.adapter_down[0].weight)
            if completed_steps >= max_train_steps:
                break
            metric_list[epoch]['train_loss'].append(loss.item())
            if (step+1) % 50 == 0 or step==len(train_dataloader)-1:
                train_loss = metric_list[epoch]['train_loss']
                train_loss = np.mean(train_loss)
                logger.info({'train loss': train_loss})
                val_metrics = val(args, model, tokenizer, val_dataloader, compute_metric)
                for m, s in val_metrics.items():
                    metric_list[epoch][m] = s
    model.save_adapter(args.task_name+'/outs', args.task_name)

def val(args, model, tokenizer, val_dataloader, compute_metric):
    model.eval()
    model.set_active_adapters(args.task_name)
    loss_list = []
    gt_list = []
    pred_list = []
    for step, batch in enumerate(val_dataloader):
        with torch.no_grad():
            loss = model(input_ids=batch['source_ids'].to(args.device),
                    attention_mask=batch['source_mask'].to(args.device),
                    labels=batch['target_ids'].to(args.device),
                    decoder_attention_mask=batch['target_mask'].to(args.device)
                    ).loss
            loss_list.append(loss.item())
            outputs = model.generate(
                input_ids=batch['source_ids'].to(args.device),
                attention_mask=batch["source_mask"].to(args.device)
            )
            pred = outputs.cpu()
            pred = tokenizer.batch_decode(pred, skip_special_tokens=True)
            gt = batch['target_ids']
            gt = tokenizer.batch_decode(gt, skip_special_tokens=True)
            for o1, o2 in zip(pred, gt):
                pred_list.append(o1)
                gt_list.append(o2)
    metrics = compute_metric(pred_list, gt_list)
    val_loss = np.mean(loss_list)
    metrics['val loss'] = val_loss
    for m, s in metrics.items():
        logger.info({m: s})
    return metrics
```

for print weight, i just print model.encoder.block[0].layer[1].adapters.rotten_tomatoes.adapter_down[0].weight, and the weight remains the same as follows:
```
Parameter containing:
tensor([[ 0.0255, -0.0217,  0.0376,  ...,  0.0119, -0.0123,  0.0187],
        [-0.0100,  0.0130,  0.0084,  ..., -0.0292,  0.0155,  0.0030],
        [ 0.0117,  0.0085,  0.0095,  ...,  0.0054, -0.0174, -0.0102],
        ...,
        [ 0.0165,  0.0016,  0.0241,  ...,  0.0056, -0.0019, -0.0012],
        [-0.0283, -0.0158,  0.0150,  ...,  0.0175,  0.0120,  0.0082],
        [ 0.0177, -0.0297, -0.0010,  ..., -0.0010, -0.0058,  0.0154]],
       device='cuda:0', requires_grad=True)
```
I train adapter and T5 on rotten tomatoes in this case but fine tuning the T5 only is working,
so I really can't figure out why.
Sincerely 

cc: @calpt @AmirAktify  sorry for the bothering

**Comment by hSterz** [comment]
Hi, it looks like the optimizer is missing the adapter parameters. Try, to first add the adapter to the model with `model.add_adapter(args.task_name, config=adapter_config)` and then get the parameters for the optimizer. Otherwise, the optimizer is not updating the adapter weights. I hope this helps.

**Comment by mistel1225** [comment]
> 

Hi, it's very helpful and the issue is solved, thanks you so much!
